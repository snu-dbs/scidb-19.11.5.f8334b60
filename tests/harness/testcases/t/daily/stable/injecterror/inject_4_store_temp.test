--setup
--start-query-logging
load_library('misc')

--igdata "store(build (<v : int64> [I=0:60,1,0], I), MANY_CHUNKS)"

create temp array ARRAY1 <v:int64> [I=0:100,1,0]
create temp array ARRAY2 <v:int64> [I=0:100,1,0]
create      array ARRAY3 <v:int64> [I=0:100,1,0]

--igdata "store (build (ARRAY1, I), ARRAY1)"
--igdata "store (build (ARRAY2,-I), ARRAY2)"
--igdata "store (build (ARRAY3,-I), ARRAY3)"

--test
# injectError (instanceid, errorid)
# Error_IDs for inject error:
# 1 - a specific error injected into the Storage::writeChunk code path which is triggered at the end of writing the first chunk
# 2 - a specific error injected into the ReplicationManager::sendItem code path which is triggered when the ReplicationManager is trying to send a chunk to a remote node
# 3 - a specific error injected into the ReplicationManager::wait code path which is triggered when the ReplicationManager is trying to wait until more buffer/queue space is available in NetworkManager. This path is often taken when replication network flow control is pushing back. 
# 4 - a generic error to be interpreted by a specific operator. store(),redim_store(),sg(),rename() error out after they have done 99% of the work (99% because a coordinator will error out before creating a new version in the catalog). remove() errors out before it does any work.

##
## ERROR4
##

# show the statistics of each
aggregate(ARRAY1, SUM(v), MIN(v), MAX(v))
aggregate(ARRAY2, SUM(v), MIN(v), MAX(v))
aggregate(ARRAY3, SUM(v), MIN(v), MAX(v))

# for HASHED distribution the chunks necessarily on every instance, but are always on the coordinator
# so we inject the error on instance 0 to be sure of gettings an error
--echo This test will fail if there are no chunks on the coordinator, so the next line should be zero.
--set-format csv
aggregate(apply(filter(list('chunk map'), nelem>0), instance, inst), min(instance) as inst)
--reset-format

# inject error into a store(X,Y)
--igdata "apply(MANY_CHUNKS, inject_result, injectError (0, 4))"
--error --code=scidb::SCIDB_SE_INJECTED_ERROR::SCIDB_LE_INJECTED_ERROR "store(build(MANY_CHUNKS,0),MANY_CHUNKS2)"

# inject error into a store(X,X)
--igdata "apply(MANY_CHUNKS, inject_result, injectError (0, 4))"
--error --code=scidb::SCIDB_SE_INJECTED_ERROR::SCIDB_LE_INJECTED_ERROR "store(build(MANY_CHUNKS,0),MANY_CHUNKS)"

# check sums of ARRAY{1,2}
aggregate(filter(apply(join(ARRAY1,ARRAY2),attrSum,ARRAY1.v + ARRAY2.v), attrSum<>0),count(*))

# inject error into a write query
--igdata "apply(MANY_CHUNKS, inject_result, injectError (0, 4))"
--error --code=scidb::SCIDB_SE_INJECTED_ERROR::SCIDB_LE_INJECTED_ERROR "store(ARRAY3,ARRAY1)"

# re-check sums of ARRAY{1,2}, that they were not altered
aggregate(filter(apply(join(ARRAY1,ARRAY2),attrSum,ARRAY1.v + ARRAY2.v), attrSum<>0),count(*))

# re-show the statistics of each
aggregate(ARRAY1, SUM(v), MIN(v), MAX(v))
aggregate(ARRAY2, SUM(v), MIN(v), MAX(v))
aggregate(ARRAY3, SUM(v), MIN(v), MAX(v))


--cleanup
remove(MANY_CHUNKS)
remove(ARRAY1)
remove(ARRAY2)
remove(ARRAY3)
