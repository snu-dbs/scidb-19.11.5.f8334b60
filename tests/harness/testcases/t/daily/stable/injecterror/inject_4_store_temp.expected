SCIDB QUERY : <load_library('misc')>
Query was executed successfully

SCIDB QUERY : <store(build (<v : int64> [I=0:60,1,0], I), MANY_CHUNKS)>
[Query was executed successfully, ignoring data output by this query.]

SCIDB QUERY : <create temp array ARRAY1 <v:int64> [I=0:100,1,0]>
Query was executed successfully

SCIDB QUERY : <create temp array ARRAY2 <v:int64> [I=0:100,1,0]>
Query was executed successfully

SCIDB QUERY : <create      array ARRAY3 <v:int64> [I=0:100,1,0]>
Query was executed successfully

SCIDB QUERY : <store (build (ARRAY1, I), ARRAY1)>
[Query was executed successfully, ignoring data output by this query.]

SCIDB QUERY : <store (build (ARRAY2,-I), ARRAY2)>
[Query was executed successfully, ignoring data output by this query.]

SCIDB QUERY : <store (build (ARRAY3,-I), ARRAY3)>
[Query was executed successfully, ignoring data output by this query.]

SCIDB QUERY : <aggregate(ARRAY1, SUM(v), MIN(v), MAX(v))>
{i} v_sum,v_min,v_max
{0} 5050,0,100

SCIDB QUERY : <aggregate(ARRAY2, SUM(v), MIN(v), MAX(v))>
{i} v_sum,v_min,v_max
{0} -5050,-100,0

SCIDB QUERY : <aggregate(ARRAY3, SUM(v), MIN(v), MAX(v))>
{i} v_sum,v_min,v_max
{0} -5050,-100,0

This test will fail if there are no chunks on the coordinator, so the next line should be zero.
SCIDB QUERY : <aggregate(apply(filter(list('chunk map'), nelem>0), instance, inst), min(instance) as inst)>
0

SCIDB QUERY : <apply(MANY_CHUNKS, inject_result, injectError (0, 4))>
[Query was executed successfully, ignoring data output by this query.]

SCIDB QUERY : <store(build(MANY_CHUNKS,0),MANY_CHUNKS2)>
[An error expected at this place for the query "store(build(MANY_CHUNKS,0),MANY_CHUNKS2)". And it failed with error code = scidb::SCIDB_SE_INJECTED_ERROR::SCIDB_LE_INJECTED_ERROR. Expected error code = scidb::SCIDB_SE_INJECTED_ERROR::SCIDB_LE_INJECTED_ERROR.]

SCIDB QUERY : <apply(MANY_CHUNKS, inject_result, injectError (0, 4))>
[Query was executed successfully, ignoring data output by this query.]

SCIDB QUERY : <store(build(MANY_CHUNKS,0),MANY_CHUNKS)>
[An error expected at this place for the query "store(build(MANY_CHUNKS,0),MANY_CHUNKS)". And it failed with error code = scidb::SCIDB_SE_INJECTED_ERROR::SCIDB_LE_INJECTED_ERROR. Expected error code = scidb::SCIDB_SE_INJECTED_ERROR::SCIDB_LE_INJECTED_ERROR.]

SCIDB QUERY : <aggregate(filter(apply(join(ARRAY1,ARRAY2),attrSum,ARRAY1.v + ARRAY2.v), attrSum<>0),count(*))>
{i} count
{0} 0

SCIDB QUERY : <apply(MANY_CHUNKS, inject_result, injectError (0, 4))>
[Query was executed successfully, ignoring data output by this query.]

SCIDB QUERY : <store(ARRAY3,ARRAY1)>
[An error expected at this place for the query "store(ARRAY3,ARRAY1)". And it failed with error code = scidb::SCIDB_SE_INJECTED_ERROR::SCIDB_LE_INJECTED_ERROR. Expected error code = scidb::SCIDB_SE_INJECTED_ERROR::SCIDB_LE_INJECTED_ERROR.]

SCIDB QUERY : <aggregate(filter(apply(join(ARRAY1,ARRAY2),attrSum,ARRAY1.v + ARRAY2.v), attrSum<>0),count(*))>
{i} count
{0} 0

SCIDB QUERY : <aggregate(ARRAY1, SUM(v), MIN(v), MAX(v))>
{i} v_sum,v_min,v_max
{0} 5050,0,100

SCIDB QUERY : <aggregate(ARRAY2, SUM(v), MIN(v), MAX(v))>
{i} v_sum,v_min,v_max
{0} -5050,-100,0

SCIDB QUERY : <aggregate(ARRAY3, SUM(v), MIN(v), MAX(v))>
{i} v_sum,v_min,v_max
{0} -5050,-100,0

SCIDB QUERY : <remove(MANY_CHUNKS)>
Query was executed successfully

SCIDB QUERY : <remove(ARRAY1)>
Query was executed successfully

SCIDB QUERY : <remove(ARRAY2)>
Query was executed successfully

SCIDB QUERY : <remove(ARRAY3)>
Query was executed successfully

